### PR TITLE
Fix DOI Url parsing

### LINKED ISSUE #10648: Entry creation from DOI does not recognize URL encoding, but "get bibliographical data" does
**Update** See below. It is now about the "Entry from Id" dialog!

### JabRef version

Latest development branch build (please note build date below)

### Operating system

Windows

### Details on version and operating system

JabRef 5.12--2023-11-16--85cd85f, Windows 10 Pro 22H2

### Checked with the latest development build (copy version output from About dialog)

- [X] I made a backup of my libraries before testing the latest development version.
- [X] I have tested the latest development version and the problem persists

### Steps to reproduce the behaviour

1. Find a DOI that contains URL-encoded characters such as %3C (<) or %3E( >). My example DOI: https://doi.org/10.1175/1520-0493(2002)130%3C1913:EDAWPO%3E2.0.CO;2
2. Add a new entry from the DOI.
Result: It fails.
import_dialog

3. Add an empty entry.
4. Enter the DOI in its DOI field.
5. Use the "get bibliographical data from DOI" function.
Result: It works.
lookup

### Appendix

_No response_

### PR DESCRIPTION
**Pull request by subhramit**
This pull request closes #10648 
## The Issue
On attempting to use JabRef's "Import by ID" feature as described in the linked issue, while importing using DOI URLs or DOI numbers containing some special characters, such as [https://doi.org/10.1175/1520-0493(2002)130%3C1913:EDAWPO%3E2.0.CO;2](https://doi.org/10.1175/1520-0493(2002)130%3C1913:EDAWPO%3E2.0.CO;2), the following error was thrown:

image
as if, the DOI link itself was an invalid one. However, the link was perfectly fine.

## The bug
The bug was extremely subtle and tough to trace. On careful inspection it was detected that when the ```performSearchById(identifier)``` was being called somewhere along the flow, line 23 was causing the variable ```doi``` to be assigned to ```https://doi.org/10.1175/1520-0493(2002)130%3C1913:EDAWPO%3E2.0.CO```. That is, somehow the trailing "```;2```" was being truncated from the identifier variable (which held the correct link) and assigned to ```doi```. This resulted in an invalid DOI link, which was causing the HTTP status ```code 404``` to be thrown, resulting in an internal ```FetcherClientException```.
https://github.com/JabRef/jabref/blob/8d082798dbe8da0505dc86b5dea90792684b9a68/src/main/java/org/jabref/logic/importer/CompositeIdFetcher.java#L22-L23
Then, on investigating the underlying ```findInText(identifier)``` function in the ```DOI``` class, the following snippet was analyzed:
https://github.com/JabRef/jabref/blob/369b9a7b00cd9332fee7f1cd222bae7dc4db6c00/src/main/java/org/jabref/model/entry/identifier/DOI.java#L202-L222
It turns out that our input ```identifier = https://doi.org/10.1175/1520-0493(2002)130%3C1913:EDAWPO%3E2.0.CO;2``` was being matched (partially) with ```FIND_DOI_PATT```, which in turn contained a section that contained ```FIND_DOI_EXP```, which had the following structure:
https://github.com/JabRef/jabref/blob/369b9a7b00cd9332fee7f1cd222bae7dc4db6c00/src/main/java/org/jabref/model/entry/identifier/DOI.java#L43-L51
In line 50, we can see that the pattern forbade ```;``` to be in any part of the suffix, so everything before that in our example was being matched. Then, when the ```DOI(new DOI(matcher.group(1));``` constructor (line 208 as seen in the ```DOI.java``` snippet linked above) was being called due to the match, it was identifying the partially matched part (```https://doi.org/10.1175/1520-0493(2002)130%3C1913:EDAWPO%3E2.0.CO```) to be a valid DOI URL, and ```result``` was being assigned that. Although a partial match  in the second case was triggering the second ```if``` statement as well, the resultant matched part was ```https://doi.org/10.1175/1520-0493(2002)130```, which was not being identified as a valid DOI URL by the ```DOI(matcher.group(1))``` constructor call, and thus not overwriting ```result```. Similarly third ```if``` statement was also immaterial here due to a mismatch or a partial match resulting in an invalid DOI URL.
So, the final value of ```result``` was ```https://doi.org/10.1175/1520-0493(2002)130%3C1913:EDAWPO%3E2.0.CO```, which was an invalid DOI mistaken to be a valid one.

## The fix
The fix was relatively simpler once the investigation was done. We just had to remove the ```;``` from the forbidden characters in the expected regex structure of ```FIND_DOI_EXP```, so that our link would fully match its structure and thus the variable ```result``` would contain the correct DOI URL, without the trailing "```;2```" being truncated. ```FIND_DOI_EXP``` **had only one usage** (as per IntelliJ), **so it could be safely modified**.
A test has also been added in ```DOITest.java```.
## Result
image
The article was successfully retrieved using Import by ID.
### Mandatory checks

- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by subhramit** [comment]
Unit test failing (```DOITest.java```). Need help fixing the test.

**Comment by subhramit** [comment]
> Unit test failing (`DOITest.java`). Need help fixing the test.

Never mind, fixed.

**Comment by subhramit** [comment]
> Unit test failing (`DOITest.java`). Need help fixing the test.

Okay, seems like it is failing not because of my changes. Please take a look.

**Comment by Siedlerchr** [comment]
That's a tough one and I am not sure how often that case with ;something happens e.g. text coming after the DOI
Some readings on Regex and DOI:
It's impossible to catch all edge cases 
https://datacite.org/blog/cool-dois/
https://www.crossref.org/blog/dois-and-matching-regular-expressions/

**Comment by subhramit** [comment]
> That's a tough one and I am not sure how often that case with ;something happens e.g. text coming after the DOI Some readings on Regex and DOI: It's impossible to catch all edge cases https://datacite.org/blog/cool-dois/ https://www.crossref.org/blog/dois-and-matching-regular-expressions/

I just added a space before "something" in my last commit. This will cause the last character to be ';', which will not be matched, resulting in the doi expected.

**Comment by subhramit** [comment]
> Edit// This one is a hard one: [https://doi.org/10.1002/(sici)1099-1409(199908/10)3:6/7<672::aid-jpp192>3.0.co;2-8](https://doi.org/10.1002/(sici)1099-1409(199908/10)3:6/7%3C672::aid-jpp192%3E3.0.co;2-8)
Tested in my build, works perfectly fine.

**Comment by subhramit** [comment]
Awaiting further comments.

**Comment by Siedlerchr** [comment]
I'm okay with this we cannot capture all edge cases

**Review by Siedlerchr — APPROVED** [review]

File: src/test/java/org/jabref/model/entry/identifier/DOITest.java [hunk 1]
```diff
@@ -316,4 +316,10 @@ public void rejectMissingDividerInShortDoi() {
     public void rejectNullDoiParameter() {
         assertThrows(NullPointerException.class, () -> new DOI(null));
     }
+
+    @Test
+    public void findDoiWithSpecialCharactersInText() {
```
Comment (koppor) [inline src/test/java/org/jabref/model/entry/identifier/DOITest.java:321]: For the future: Integrate this in the exiting `@ParamterizedTest`. The method name is then the comment above the pair of `Arguments.of`. See Line 223.

Comment (subhramit) [reply src/test/java/org/jabref/model/entry/identifier/DOITest.java:321]: Had forgotten about this. Done in https://github.com/JabRef/jabref/pull/11603.
